File: Solutions/56_Merge_Intervals/56_Merge_Intervals.cs
//Quite ellegant
public class Solution {
    public int[][] Merge(int[][] intervals) {
        var res = new Stack<int[]>();
        foreach(var interval in intervals.OrderBy(x => x[0])){
            if(res.TryPeek(out var last) && last[1] >= interval[0])
                last[1] = Math.Max(last[1], interval[1]);
            else
                res.Push(interval);
        }
        
        return res.ToArray();
    }
}

//Third fastest
public class Solution {
    public int[][] Merge(int[][] intervals) {
        Array.Sort(intervals, (x, y) => x[0].CompareTo(y[0]));
        var result = new List<int[]>();
        var temp = intervals[0];                       
        
        for(int i = 1; i < intervals.Length; i++) {
            var c = intervals[i];
            if(temp[1] < c[0]) {
                result.Add(temp);
                temp = c;
                continue;
            }
            temp = new int[] {Math.Min(temp[0], c[0]), Math.Max(temp[1], c[1])};
        }
        result.Add(temp);
        return result.ToArray();
    }
}

//Second fastest
public class Solution {
    public int[][] Merge(int[][] intervals) {
        List<int[]> res = new List<int[]>();
        if (intervals.Length == 0)  return res.ToArray();
        Array.Sort(intervals, (a,b)=>(a[0].CompareTo(b[0])));
        int start = intervals[0][0], end = intervals[0][1];
        foreach(int[] ivl in intervals) {
            if (ivl[1] < end)   continue;
            if (ivl[0] > end) {
                res.Add(new int[2]{start, end});
                start = ivl[0];
                end = ivl[1];
                continue;
            }
            end = ivl[1];
        }
        res.Add(new int[2]{start, end});
        return res.ToArray();
    }
}

//Fastest
public class Solution {
    public int[][] Merge(int[][] intervals) {
        if(intervals?.Length == 0) {
            return intervals;
        }
        
        Array.Sort(intervals, new Comparer());

        List<int[]> result = new List<int[]>();
        int[] interval = intervals[0];
        for(int i = 1; i < intervals.Length; i++) {
            if(intervals[i][0] > interval[1]) {
                result.Add(interval);
                interval = intervals[i];
            }
            else {
                interval[1] = Math.Max(intervals[i][1], interval[1]);
            }
        }
        result.Add(interval);
        return result.ToArray();
    }
    
    public class Comparer : IComparer
    {
        public int Compare(object o1, object o2) {
            int[] arr1 = (int[])o1;
            int[] arr2 = (int[])o2;
            
            return arr1[0] == arr2[0] ? arr1[1].CompareTo(arr2[1]) : arr1[0].CompareTo(arr2[0]);
        }
    }
}

File: Solutions/56_Merge_Intervals/56_Merge_Intervals.go
//Third fastest
type Intervals [][]int

func (i Intervals) Less(a, b int) bool {
    return i[a][0] < i[b][0]
}

func (i Intervals) Swap(a, b int) {
    i[a], i[b] = i[b], i[a]
}

func (i Intervals) Len() int {
    return len(i)
}

func merge(intervals [][]int) [][]int {
    var i Intervals
    i = intervals
    sort.Sort(i)
    
    i = append(i, []int{math.MaxInt32, -1})
    result := [][]int{}
    end := i[0][1]
    start := i[0][0]
    for _, r := range i {
        if r[0] > end {
            result = append(result, []int{start, end})   
            start = r[0]
        }
        
        end = max(end, r[1])
    }
    
    return result
}

func max(a,b int) int {
    if a > b {
        return a
    }
    
    return b
}

//Second fastest
func merge(intervalsCop [][]int) [][]int {
    out := make([][]int,0)
    sort.Slice(intervalsCop, func(i, k int) bool {
        return intervalsCop[i][0] < intervalsCop[k][0]
    })    
    start := false
    for i:=0; i<= len(intervalsCop)-1;i++{
        if intervalsCop[i] == nil{
            continue
        }
        if ! start {
            out  = append(out, intervalsCop[i]) 
            start = true
            continue
        }

        if out[len(out)-1][1]>= intervalsCop[i][0] && out[len(out)-1][1]<= intervalsCop[i][1]{
                out[len(out)-1] = []int{
                    out[len(out)-1][0],
                    intervalsCop[i][1],
                }
        }else if   out[len(out)-1][1] < intervalsCop[i][0]{
            out = append(out, intervalsCop[i])
        }
        
    }
    return out
    
}

//Fastest
type INTERVALS [][]int

func (I INTERVALS) Less(i, j int) bool {
	return I[i][0] < I[j][0]
}
func (I INTERVALS) Swap(i, j int) {
	I[i], I[j] = I[j], I[i]
}
func (I INTERVALS) Len() int {
	return len(I)
}

func merge(intervals [][]int) [][]int {
	sortByBIdx(intervals)
	var ans [][]int
	for _, v := range intervals {
		if len(ans) == 0 {
			ans = append(ans, v)
			continue
		}
		lastIdx := len(ans) - 1
		if v[0] > ans[lastIdx][1] {
			ans = append(ans, v)
			continue
		}

		if v[1] > ans[lastIdx][1] {
			ans[lastIdx][1] = v[1]
			continue
		}
	}

	return ans
}

func sortByBIdx(intervals [][]int) [][]int {
	sort.Sort(INTERVALS(intervals))
	return intervals
}